core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java

```diff
@@ -644,7 +644,18 @@ public synchronized OutputSource createOutputSource(Set<TaskId> selectedTasks)
                         }
                         return getAllSourceHandles(exchange.getSourceHandles());
                     }, executor)
-                    .transform(exchangeManager::createSource, executor);
+                    .transform(handles -> {
+                        ExchangeSource source = exchangeManager.createSource();
+                        try {
+                            source.addSourceHandles(handles);
+                            source.noMoreSourceHandles();
+                            return source;
+                        }
+                        catch (Throwable t) {
+                            source.close();
```

Comment:
nit: you may catch exception from `source.close()` and add those to `t` as suppressed.